kyo-data/shared/src/main/scala/kyo/SafeClassTag.scala

```diff
@@ -0,0 +1,113 @@
+
+object SafeClassTag:
+    sealed trait Element
+    case class Union(elements: List[SafeClassTag[Any]])        extends Element
+    case class Intersection(elements: List[SafeClassTag[Any]]) extends Element
+
+    sealed trait Primitive extends Element
+
+    case object IntTag     extends Primitive
+    case object LongTag    extends Primitive
+    case object DoubleTag  extends Primitive
+    case object FloatTag   extends Primitive
+    case object ByteTag    extends Primitive
+    case object ShortTag   extends Primitive
+    case object CharTag    extends Primitive
+    case object BooleanTag extends Primitive
+    case object UnitTag    extends Element
+    case object AnyValTag  extends Element
+    case object NothingTag extends Element
+
+    inline given apply[A]: SafeClassTag[A] = ${ SafeClassTagMacro.derive[A] }
+
+    extension [A](self: SafeClassTag[A])
+
+        def accepts(value: Any): Boolean =
+            !isNull(value) && {
+                given CanEqual[Any, Any] = CanEqual.derived
+                self match
+                    case Union(elements)        => elements.exists(_.accepts(value))
+                    case Intersection(elements) => elements.forall(_.accepts(value))
+                    case NothingTag             => false
+                    case UnitTag                => value == ()
+                    case AnyValTag =>
+                        value match
+                            case (_: Int | _: Long | _: Double | _: Float | _: Byte | _: Short | _: Char | _: Boolean) =>
+                                true
+                            case _ =>
+                                val cls = value.getClass
+                                classOf[AnyVal].isAssignableFrom(cls) && (cls ne classOf[AnyVal])
+                    case _: Primitive =>
+                        value match
+                            case _: Int     => self eq IntTag
+                            case _: Long    => self eq LongTag
+                            case _: Double  => self eq DoubleTag
+                            case _: Float   => self eq FloatTag
+                            case _: Byte    => self eq ByteTag
+                            case _: Short   => self eq ShortTag
+                            case _: Char    => self eq CharTag
+                            case _: Boolean => self eq BooleanTag
+                            case _          => false
+                    case self: Class[?] => self.isInstance(value)
+                end match
+            }
+        end accepts
+
+        private def isCustomAnyVal(value: Any): Boolean =
+            if isNull(value) then false
+            else
+                val clazz       = value.getClass
+                val anyValClass = classOf[AnyVal]
+                anyValClass.isAssignableFrom(clazz) && (clazz ne anyValClass)
+        end isCustomAnyVal
+
+        def unapply(value: Any): Maybe.Ops[A] =
```

Comment:
I had put `SafeClassTag` in `kyo-tag` initially but moved to `kyo-data` to be able to use `Maybe.Ops` to avoid allocations. I'm planning to follow up with another PR merging `kyo-tag` into `kyo-data`. I don't think there's much benefit in having them separate.